crates/stackable-operator/src/validation.rs

```diff
@@ -26,121 +27,202 @@ const RFC_1123_SUBDOMAIN_MAX_LENGTH: usize = 253;
+        }
+        .into())
+    }
+}
+
+/// Returns [`Ok`] if *all* validations are [`Ok`], otherwise returns all errors.
+fn validate_all(validations: impl IntoIterator<Item = Result<(), Error>>) -> Result {
+    let errors = validations
+        .into_iter()
+        .filter_map(|res| res.err())
+        .collect::<Vec<_>>();
     if errors.is_empty() {
         Ok(())
     } else {
-        Err(errors)
+        Err(Errors(errors))
     }
 }
 
+/// Tests for a string that conforms to the definition of a subdomain in DNS (RFC 1123).
+pub fn is_rfc_1123_subdomain(value: &str) -> Result {
+    validate_all([
+        validate_str_length(value, RFC_1123_SUBDOMAIN_MAX_LENGTH),
+        validate_str_regex(
+            value,
+            &RFC_1123_SUBDOMAIN_REGEX,
+            RFC_1123_SUBDOMAIN_ERROR_MSG,
+            &["example.com"],
+        ),
+    ])
+}
+
 /// Tests for a string that conforms to the definition of a label in DNS (RFC 1123).
 /// Maximum label length supported by k8s is 63 characters (minimum required).
-pub fn is_rfc_1123_label(value: &str) -> Result<(), Vec<String>> {
-    let mut errors = vec![];
-    if value.len() > RFC_1123_LABEL_MAX_LENGTH {
-        errors.push(max_len_error(RFC_1123_LABEL_MAX_LENGTH))
-    }
-
-    // Regex is identical to RFC 1123 subdomain
-    if !RFC_1123_SUBDOMAIN_REGEX.is_match(value) {
-        errors.push(regex_error(
+pub fn is_rfc_1123_label(value: &str) -> Result {
+    validate_all([
+        validate_str_length(value, RFC_1123_LABEL_MAX_LENGTH),
+        validate_str_regex(
+            value,
+            &RFC_1123_LABEL_REGEX,
             RFC_1123_LABEL_ERROR_MSG,
-            RFC_1123_SUBDOMAIN_FMT,
             &["example-label", "1-label-1"],
-        ))
-    }
-
-    if errors.is_empty() {
-        Ok(())
-    } else {
-        Err(errors)
-    }
+        ),
+    ])
 }
 
 /// Tests for a string that conforms to the definition of a label in DNS (RFC 1035).
-pub fn is_rfc_1035_label(value: &str) -> Result<(), Vec<String>> {
-    let mut errors = vec![];
-    if value.len() > RFC_1035_LABEL_MAX_LENGTH {
-        errors.push(max_len_error(RFC_1035_LABEL_MAX_LENGTH))
-    }
-
-    if !RFC_1035_LABEL_REGEX.is_match(value) {
-        errors.push(regex_error(
-            RFC_1035_LABEL_ERR_MSG,
-            RFC_1035_LABEL_FMT,
+pub fn is_rfc_1035_label(value: &str) -> Result {
+    validate_all([
+        validate_str_length(value, RFC_1035_LABEL_MAX_LENGTH),
+        validate_str_regex(
+            value,
+            &RFC_1035_LABEL_REGEX,
+            RFC_1035_LABEL_ERROR_MSG,
             &["my-name", "abc-123"],
-        ))
-    }
+        ),
+    ])
+}
 
-    if errors.is_empty() {
-        Ok(())
-    } else {
-        Err(errors)
-    }
+/// Tests whether a string looks like a reasonable Kerberos realm name.
+///
+/// This check is much stricter than krb5's own validation,
+pub fn is_kerberos_realm_name(value: &str) -> Result {
+    validate_all([validate_str_regex(
```

Comment:
question: Is it worth using the `validate_all` function here, if we only provide a single validation function?

Reply:
It ensures that we still return the `Errors` rather than `Error`, and gives us room to add additional validations later on.

Reply:
Sounds reasonable, thx!